Generate initial data

kkuna opened the pull request:
--------------------------------------------------------------------
* Added seed data
* Created handlers to store/get data from AsyncStorage
* Added sagas to handle identity flow
* Added functions to normalize/denormalize seed AsyncStorage data
* Added selectors for the identity data
* Added reducer for the identity data
* Persist identity that is currently active
* Add new identity to storage
--------------------------------------------------------------------

kkuna committed:
CHECKPOINT / Generate initial data

--------------------------------------------------------------------
 MAIN
 * Added seed data
 * Created handlers to store/get data from AsyncStorage
 * Added sagas to handle identity flow
 * Added functions to normalize/denormalize seed AsyncStorage data
 * Added selectors for the identity data
 * Added reducer for the identity data
--------------------------------------------------------------------

Ant-Nazif committed:
CHECPOINT/Fixes claims

kkuna committed:
CHECKPOINT / Persist identities in storage

--------------------------------------------------------------------
 MAIN
 * Persist identity that is currently active
 * Add new identity to storage
--------------------------------------------------------------------

kkuna committed:
CHECKPOINT / Generate data for identity

--------------------------------------------------------------------
 MAIN
 * Generate data when user changes identity
--------------------------------------------------------------------
 OTHER
 * Add identity view cleanup
--------------------------------------------------------------------

hanzah commented on src/containers/Identity/AddIdentity/Container.js:
```diff
@@ -0,0 +1,23 @@
+
+const mapStateToProps = (state) => ({
+  identities: selectIdentities(state),
+  activeIdentity: selectActiveIdentityId(state),
```
Work with models, not ids.
Feel free to pass in the whole identity object, so this can be selectActiveIdentity.

hanzah commented on src/containers/Identity/AddIdentity/View.js:
```diff
@@ -1,37 +1,27 @@
-            setName(null);
-        }
+        if (!name) return
+        actions.addNewIdentityName(name.includes('@') ? name : `${name}@`);
```
Create a generic action updateIdentity, where you can pass in an object with the properties that you would like to update

hanzah commented on src/containers/Identity/AddIdentity/View.js:
```diff
@@ -1,37 +1,27 @@
+        if (activeIdentity) {
+            navigation.navigate('Identity', { selectedScreen: "Identity"});
+        }
+            actions.changeActiveIdentity(identity.get('id', ''), identity)
```
Why does this action take an id?
If you are naming it changeActiveIdentity the saga can know what the active identity is. The view doesn't have to care about it.

hanzah commented on src/containers/Identity/Home/AttestationDetails/Container.js:
```diff
@@ -0,0 +1,21 @@
+
+const mapStateToProps = (state) => ({
+  attestation: selectActiveAttestation(state),
+  activeAttestationId: selectActiveAttestationId(state),
```
Same comments as above.

hanzah commented on src/containers/Identity/Home/AttestationDetails/View.js:
```diff
@@ -1,24 +1,23 @@
+    const { attestation, activeAttestationId, actions: { toggleAttestationPin }} = props;
+
+    const toggleSwitch = (value) => {
+        toggleAttestationPin(attestation.getIn([activeAttestationId, 'id'], ''), value)
```
Move knowledge of activeAttention to saga.

hanzah commented on src/containers/Identity/Home/View.js:
```diff
@@ -1,31 +1,78 @@
+            <Text style={styles.textBadge}>Pinned attestations</Text>
+            <ScrollView>
+                <View>
+                    {pinnedAttestations.keySeq().map(attestationKey => (
```
Another thing, if you don't need this kind of objects as an immutableMap, for lookups for example, transfer them into immutable lists in the selector, so you can simply map over them in the view. Adjust your data to the view.

hanzah commented on src/containers/Identity/Home/View.js:
```diff
@@ -1,31 +1,78 @@
-export default Home;
\ No newline at end of file
+
+const mapStateToProps = (state) => ({
```
Think about refactoring into a separate file, like we do for the components in general.

hanzah commented on src/containers/Identity/PersonalInfo/ClaimDetails/View.js:
```diff
@@ -0,0 +1,99 @@
+    const newData = attestationsData.filter((item) => {
+      const itemData = item.get("id", "").toUpperCase();
+      const textData = value.toUpperCase();
+      return itemData.indexOf(textData) > -1;
```
You can use includes. It's a bit more readable.

hanzah commented on src/containers/Identity/PersonalInfo/ClaimDetails/View.js:
```diff
@@ -0,0 +1,99 @@
+        </View>
+        <View style={{ paddingVertical: 50 }}>
+          {childClaims.size > 0 ?
+            <Text style={{ paddingHorizontal: 16 }}>Child Claims</Text> : null}
```
Move styles into styles file.

hanzah commented on src/containers/Identity/PersonalInfo/ClaimDetails/View.js:
```diff
@@ -0,0 +1,99 @@
+            />
+          ))
+          }
+          {parentClaims.size > 0 ?
```
This can be a separate component, that you can reuse for childClaims and parentClaims. They are fairly the same here.

hanzah commented on src/containers/Identity/PersonalInfo/View.js:
```diff
@@ -1,21 +1,25 @@
+const PersonalInfo = (props) => {
+    const { claimCategories, navigation } = props;
+    const [categories, setCategories ] = useState(claimCategories);
+    const [value,setValue] = useState('');
```
Add space in front of comma. Check your linter.

hanzah commented on src/sagas/identity.js:
```diff
@@ -0,0 +1,136 @@
+}
+
+function* handleStoreIdentities() {
+  const storedIdentities = yield call(getIdentities);
```
This is doing the same thing as handleSeedData? 
What is the reasoning behind it?

hanzah commented on src/reducers/identity.js:
```diff
@@ -0,0 +1,106 @@
+    case SET_ACTIVE_IDENTITY:
+      return state.setIn(
+        ["personalInformation", "activeIdentity"],
+        fromJS(action.payload.identity)
```
If you are normalizing your state, the activeIdentity should be an id, not an object. At the same time, add it to the defaultState.
You will then fetch it from the identities -> byId map when you need it in the selector.

hanzah commented on src/reducers/identity.js:
```diff
@@ -0,0 +1,106 @@
+        ["personalInformation", "activeClaimCategoryId"],
+        fromJS(action.payload.activeClaimCategoryId)
+      );
+      case SET_ACTIVE_CLAIM:
```
Formatting and indentation.

hanzah commented on src/containers/Login/Login.js:
```diff
@@ -84,7 +85,7 @@ class Login extends Component {
           activateChainLifecycle(coinObj.id);
         });
-
+        this.props.dispatch(requestSeedData());
```
Seed data should only be populated into async storage once when you run the app the first time.
You need to distinguish the 2 events
seedData - This happend when the user loads the app for the first time and you should store all the constants into the application
loadData - This event happens each time the user runs the app or logs in depending if the seed data is user specific. For now you can leave it in this place.
So, create a seedData event that will check if asyncStorage is empty. If it is, store your constants in it. The app should show a loading spinner while doing so.
After the seed data is stored, you can continue your flow where you will populate your store with the data from async storage.

hanzah commented on src/sagas/identity.js:
```diff
@@ -0,0 +1,136 @@
+
+function* handleStoreIdentities() {
+  const storedIdentities = yield call(getIdentities);
+  const identities = yield call(normalizedIdentities, storedIdentities);
```
Don't name your function with a noun. It represents an action, so name it normalizeSomething for example.

hanzah commented on src/sagas/identity.js:
```diff
@@ -0,0 +1,136 @@
+function* handleSetActiveIdentity() {
+  const selectedIdentityId = yield select(selectActiveIdentityId);
+    yield all([
+      call(storeSeedClaimCategories, selectedIdentityId),
```
Seeding data needs to be independent of the logic you use to populate your store on a normal startup

Here:

start app first time
dispatch(appSetup())
takeLatest('APP_SETUP', handleAppSetup)
handleAppSetup = () => {
    const dataSeeded = yield call(fetchPropertyFromAsyncStoreThatIndicatesDataIsSeeded)
    if !dataSeeded {
       return put(seedData)
    }
    yield put(appSetupCompleted)
}

takeLatest('SEED_DATA', handleSeedData)

handleSeedData = () => {
    take all your consts and drop them into async storage
    this is async so wait for it to complete, then dispatch
    yield put(appSetupCompleted)
}

Now continue to login flow
In the login flow, you don't care about the seed data anymore. Now you just need to load the data from asyncStorage, and store it into your store.

hanzah commented on src/selectors/identity.js:
```diff
@@ -0,0 +1,74 @@
+export const selectIdentityObj = (state) => state.identity.getIn(["personalInformation", "identities"], IMap())
+
+export const selectActiveIdentity = (state) =>
+  state.identity.getIn(["personalInformation", "activeIdentity"], IMap());
```
So this logic should use createSelector to 
const selectActiveCategory = createSelector(
  (state) => state.personalInformation.activeIdentity,
  (state) => state.personalInformation.identities 
  (activeIdentity, identities) => identities.byId[activeIdentity])
);
Adopt just for immutable syntax. The intention is obvious. The same for other stuff. You store references to the entities tables in your store and then cache/memoize them in your selectors.

hanzah commented on src/utils/asyncStore/identityStorage.js:
```diff
@@ -0,0 +1,134 @@
+import claimCategories from '../InitialData/ClaimCategory';
+import attestations from '../InitialData/Attestation';
+
+export const storeSeedIdentities = () => {
```
Use async await

Ant-Nazif committed:
CHECKPOINT/Merge fix

Ant-Nazif committed:
CHECKPOINT/Fixes

Ant-Nazif committed:
CHECKPOINT/Fixes add identity

kkuna committed:
COMPLETED / Code cleanup

--------------------------------------------------------------------
 MAIN
 * Made changes according to PR review comments
 * Added eslint configuration
--------------------------------------------------------------------

hanzah commented on src/containers/Identity/View.js:
```diff
@@ -1,62 +1,57 @@
+  const [activeTab, setActiveTab] = useState(tabs[0]);
+
+  const selectedTab = navigation.state.params.selectedScreen;
+  const tabIndex = Object.keys(tabs).findIndex((tabKey) => tabs[tabKey].label.toLowerCase() === selectedTab.toLowerCase());
```
You need to put this in a useMemo

hanzah replied:
Tabs is a constant as well. So you can optimize the lookup by having
tabsByKey = {key1: tab1, key2: tab2} and you don't need the find at all.
